src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIPhysicalHandle.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.Data.SqlClient.SNI
+{
+    internal abstract class SNIPhysicalHandle : SNIHandle
+    {
+        protected const int DefaultPoolSize = 4;
```

Comment:
Does changing this pool size make considerable difference and should we allow it to be configurable for users who are interested in playing with different sizes for performance?

Reply:
The pool size is per-connection so it is likely to stay within these bounds. 4 comes from the count of 1 sending, 1 cancel, 1 receive and 1 building packet. That's the max I thought could be needed. I think it's unlikely that most connections will use more than 2 but I wasn't trying to optimize the pool just get it in and working.

Reply:
I was thinking case of large data transfer. Will increasing packet pool size give us benefit in reading large datasets?

Reply:
Slightly surprisingly no. The packets and the buffers that they use are entirely recycled. When a packet is received the contents is copied out into the TdsParserStateObject._inBuff, if the inbuff is needed as part of a snapshot a new inbuff is created. So there are effectively two closed loops exchanging data, this work only changes the SNI side of the loop and in future I'll see if I can improve the tdsparser side. Trying to rework the async snapshot mechanism makes me a bit nervous because it's considerably more subtle than the SNI side.